### PR TITLE
V2 CLI

### PR DESCRIPTION
**Pull request by nicktate**
* Complete overhaul of CLI system

@normanjoyner

### CODE REVIEW

File: commands/app.js [hunk 1]
```diff
@@ -1,676 +1,397 @@
...
+                'CONSTRAINTS',
+                'TAGS'
+            ];
+
+            const env_vars = _.map(app.env_vars, (v, k) => {
+                return `${chalk.gray(k)}: ${v}`;
+            });
+
+            const volumes = _.map(app.volumes, (vol) => {
+                let def = `${vol.host}:${vol.container}`;
```
Comment (mikestoltz) [inline commands/app.js]: `return `${vol.host}:${vol.container}${vol.propogation ? `:${vol.propogation}` : ''}`;`

File: commands/app.js [hunk 2]
```diff
@@ -1,676 +1,397 @@
...
+            app.tags = flatten(app.tags);
+            const to_remove_tag = _.keys(flatten(_.pickBy(options.tags, val => val !== false && !val)));
+            to_remove_tag.forEach((key) => {
+                delete app.tags[key];
+            });
+            const to_add_tag = flatten(_.pickBy(options.tags, val => val));
+            options.tags = unflatten(_.merge(app.tags, to_add_tag));
+
+            // volumes
+            options.volumes.forEach((vol) => {
```
Comment (mikestoltz) [inline commands/app.js]: _.forEach

File: commands/app.js [hunk 3]
```diff
@@ -1,676 +1,397 @@
...
+    description: 'Scale up application containers.',
+    options: {
+        count: {
+            description: 'Number of containers to scale application up by.',
+            alias: 'c',
+            default: 1
+        }
+    },
+    callback: (argv) => {
+        return request.post(`applications/${argv.app_name}/containers`, { qs: { count: argv.count } }, null, (err, response) => {
```
Comment (mikestoltz) [inline commands/app.js]: Should the qs be in a child object here? I thought the second arg was the raw qs object

Comment (mikestoltz) [inline commands/app.js]: should the qs be in a child qs object for non cloud proxy requests?

File: commands/app.js [hunk 4]
```diff
@@ -1,676 +1,397 @@
...
+    description: 'Scale down application containers.',
+    options: {
+        count: {
+            description: 'Number of containers to scale application down by.',
+            alias: 'c',
+            default: 1
+        }
+    },
+    callback: (argv) => {
+        return request.delete(`applications/${argv.app_name}/containers`, { qs: { count: argv.count } }, (err, response) => {
```
Comment (mikestoltz) [inline commands/app.js]: here as well

File: commands/app.js [hunk 5]
```diff
@@ -1,676 +1,397 @@
...
-            }
-        }
+    if(_.has(options, 'volume')) {
+        options.volumes = utils.parse_volumes(options.volume);
+        delete options.volume;
+        delete options.b;
+    }
+
+    if(_.has(options, 'env-var')) {
+        options.env_vars = utils.parse_tags(options['env-var']);
```
Comment (mikestoltz) [inline commands/app.js]: parse_env_vars?

File: commands/app.js (hunk 1 above)
Comment (nicktate) [inline commands/app.js]: no thanks

File: commands/client-plugin.js [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+        if (response.statusCode !== 200) {
+            return callback('Failed to search for plugins.');
+        }
+
+        return callback(null, response.body);
+    });
+}
+
+module.exports.commands.push({
+    name: 'search [name]',
```
Comment (mikestoltz) [inline commands/client-plugin.js]: sometimes you use <> to represent variable input and sometimes you use [], we should make it consistent

File: commands/app.js [hunk 7]
```diff
@@ -1,676 +1,391 @@
...
-var colors = require("colors");
-
-var protocols = {
-    http: require("http"),
-    https: require("https")
-}
+'use strict';
 
-module.exports = {
+const request = require('../lib/request');
```
Comment (normanjoyner) [inline commands/app.js]: alphabetize

File: commands/app.js [hunk 8]
```diff
@@ -1,676 +1,391 @@
...
-                }
+module.exports = {
+    name: 'app',
+    description: 'List and manipulate applications running on the cluster specified by the client remote.',
+    commands: []
+};
+
+module.exports.commands.push({
+    name: 'list',
+    description: 'List applications on the cluster',
```
Comment (normanjoyner) [inline commands/app.js]: punctuation

File: commands/app.js [hunk 9]
```diff
@@ -1,676 +1,391 @@
...
-                }
+            const output = Table.createTable(headers, data);
+            return console.info(output);
+        });
+    }
+});
+
+module.exports.commands.push({
+    name: 'show <app_name>',
+    description: 'Show detailed information for specified application',
```
Comment (normanjoyner) [inline commands/app.js]: punctuation

File: commands/app.js [hunk 10]
```diff
@@ -1,676 +1,391 @@
...
-                        process.exit(1);
-                    }
-                    else if(response.statusCode != 201){
-                        process.stderr.write(response.body.error);
-                        process.exit(1);
-                    }
-                    else
-                        process.stdout.write(["Successfully created application ", options.application, "!"].join(""));
-                });
+            if(response.statusCode != 200) {
```
Comment (normanjoyner) [inline commands/app.js]: `!==`

**Pull request merged**